dmff/mbar.py

```diff
@@ -0,0 +1,194 @@
+import numpy as np
+import mdtraj as md
+from pymbar import MBAR
+import jax
+import jax.numpy as jnp
+from jax import grad
+from tqdm import tqdm, trange
+
+
+class BaseState:
+    def __init__(self, temperature, name):
+        self._temperature = temperature
+        self.name = name
+
+    def calc_energy_frame(self, frame):
+        # implement this on your own
+        return 0.0
+
+    def calc_energy(self, trajectory):
+        # return beta * u
+        beta = 1. / self._temperature / 8.314 * 1000.
+        eners = []
+        for frame in tqdm(trajectory):
+            e = self.calc_energy_frame(frame)
+            eners.append(e * beta)
+        return np.array(eners)
+
+
+class Sample:
+    def __init__(self, trajectory, from_state):
+        self.trajectory = trajectory
+        self.from_state = from_state
+        self.energy_data = {}
+
+    def generate_energy(self, state_list):
+        for state in state_list:
+            if state.name not in self.energy_data:
+                self.energy_data[state.name] = np.array(
+                    [state.calc_energy(self.trajectory)])
+
+
+class MBAREstimator:
+    def __init__(self):
+        self.samples = []
+        self.states = []
+        self._mbar = None
+        self._umat = None
+        self._nk = None
+        self._full_samples = None
+
+    def add_sample(self, sample):
+        self.samples.append(sample)
+
+    def add_state(self, state):
+        self.states.append(state)
+
+    def remove_sample(self, name):
+        init_num = len(self.samples)
+        self.samples = [s for s in self.samples if s.from_state != name]
+        final_num = len(self.samples)
+        assert init_num < final_num
+
+    def remove_state(self, name):
```

Comment:
When a state is removed, it's corresponding samples should be removed too

Reply:
Done